File: trunk/java/formulario-integrado/src/formulario/integrado/business/FormularioBusiness.java
package formulario.integrado.business;

import formulario.integrado.model.Categoria;
import formulario.integrado.model.Formulario;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class FormularioBusiness extends Business<Formulario> implements IFormularioBusiness {

    private PreparedStatement ps;
    private ResultSet rs;
    private String sql;
    
    @Override
    public Formulario find(int id) throws SQLException {
        super.openConnection();
        
        Formulario formulario = null;
        this.rs = null;
        
        this.sql = "SELECT f.id, f.titulo, f.status, f.aberto, f.data "
                 + "FROM formulario f WHERE f.id = ?;";
        
        this.ps = connection.prepareStatement(this.sql);
        this.ps.setInt(1, id);
        this.rs = this.ps.executeQuery();
        
        while (rs.next()) {
            ArrayList<Categoria> categorias = new ArrayList<>();

            String query = "SELECT c.id, c.titulo, c.descricao, c.status, c.data "
                        + "FROM formulario f "
                        + "INNER JOIN formulario_categoria fc ON f.id = fc.formularioId AND f.id = ? "
                        + "INNER JOIN categoria c ON c.id = fc.categoriaId AND c.status = 1";
            
            PreparedStatement preparedStatement = connection.prepareStatement(query);
            preparedStatement.setInt(1, id);
            ResultSet resultSet = preparedStatement.executeQuery();
            
            while (resultSet.next()) {
                categorias.add(assemblyCategoria(resultSet));
            }
            
            formulario = assemblyFormulario(rs, categorias);
        }
        
        super.closeConnection();
        
        return formulario;
    }

    @Override
    public List<Formulario> show() throws SQLException {
        super.openConnection();
        
        ArrayList<Formulario> formularios = new ArrayList<>();
        this.rs = null;
        
        this.sql = "SELECT f.id, f.titulo, f.status, f.aberto, f.data "
                 + "FROM formulario f WHERE f.status = 1;";
        
        this.ps = connection.prepareStatement(this.sql);
        this.rs = this.ps.executeQuery();
        
        while (rs.next()) {
            ArrayList<Categoria> categorias = new ArrayList<>();

            String query = "SELECT c.id, c.titulo, c.descricao, c.status, c.data "
                        + "FROM formulario f "
                        + "INNER JOIN formulario_categoria fc ON f.id = fc.formularioId AND f.id = ? "
                        + "INNER JOIN categoria c ON c.id = fc.categoriaId AND c.status = 1";
            
            PreparedStatement preparedStatement = connection.prepareStatement(query);
            preparedStatement.setInt(1, rs.getInt(1));
            ResultSet resultSet = preparedStatement.executeQuery();
            
            while (resultSet.next()) {
                categorias.add(assemblyCategoria(resultSet));
            }
            
            formularios.add(assemblyFormulario(rs, categorias));
        }
        
        // verificar se possui categoria
        
        super.closeConnection();
        
        return formularios;
    }
    
    @Override
    public void moveDown(Categoria categoria) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void moveUp(Categoria categoria) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void add(Formulario formulario) throws SQLException {
        super.openConnection();

        this.sql = "INSERT INTO formulario (titulo, aberto, status, data) VALUES (?,?,?,?);";
        this.ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        
        this.ps.setString(1, formulario.getTitulo());
        this.ps.setBoolean(2, formulario.isAberto());
        this.ps.setBoolean(3, formulario.isStatus());
        this.ps.setString(4, super.getCurrentDate(formulario.getData()));
        
        this.ps.executeUpdate();
        
        this.rs = ps.getGeneratedKeys();
        this.rs.next();
        int id = this.rs.getInt(1);
                
        for (Iterator<Categoria> categorias = formulario.getCategorias().iterator(); categorias.hasNext();) {
            String query  = "INSERT INTO formulario_categoria (formularioId, categoriaId) VALUES (?,?);";
            PreparedStatement preparedStatement = connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);

            preparedStatement.setInt(1, id);
            preparedStatement.setInt(2, categorias.next().getId());
            
            preparedStatement.executeUpdate();
        }
        
        super.closeConnection();
    }

    @Override
    public void update(Formulario formulario) throws SQLException{
        super.openConnection();
        
        this.sql = "UPDATE formulario SET titulo = ?, aberto = ?, status = ? WHERE id = ?;";
        this.ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        
        this.ps.setString(1, formulario.getTitulo());
        this.ps.setBoolean(2, formulario.isAberto());
        this.ps.setBoolean(3, formulario.isStatus());
        this.ps.setInt(4, formulario.getId());
        
        this.ps.executeUpdate();
        
        super.closeConnection();
    }

    @Override
    public void remove(Formulario formulario) throws SQLException {
        super.openConnection();
        
        this.sql = "UPDATE formulario SET status = ? WHERE id = ?;";
        this.ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        
        this.ps.setBoolean(1, false);
        this.ps.setInt(2, formulario.getId());
        
        this.ps.executeUpdate();
        
        super.closeConnection();
    }
    
    /**
     * Método para popular um Formulário
     * 
     * @param ResultSet rs
     * @return Formulario
     * @throws SQLException 
     */
    private Formulario assemblyFormulario(ResultSet rs, ArrayList<Categoria> categorias) throws SQLException {
        Formulario formulario = new Formulario();
        formulario.setId(rs.getInt(1));
        formulario.setTitulo(rs.getString(2));
        formulario.setStatus(rs.getBoolean(3));
        formulario.setAberto(rs.getBoolean(4));
        formulario.setData(super.setCurrentDate(rs.getString(5)));
        
        if (categorias != null) {
            formulario.setCategorias(categorias);
        }
        
        return formulario;
    }
    
    /**
     * Método para popular Categoria
     * 
     * @param ResultSet rs
     * @return Categoria
     * @throws SQLException 
     */
    private Categoria assemblyCategoria(ResultSet resultSet) throws SQLException {
        Categoria categoria = new Categoria();
        categoria.setId(resultSet.getInt(1));
        categoria.setTitulo(resultSet.getString(2));
        categoria.setDescricao(resultSet.getString(3));
        categoria.setStatus(resultSet.getBoolean(4));        
        categoria.setData(super.setCurrentDate(resultSet.getString(5)));
        
        return categoria;
    }

    /**
     * Método para remover relacionamento de categoria com formulário
     */
    @Override
    public void remove(Formulario formulario, Categoria categoria) throws SQLException {
        super.openConnection();
        
        this.sql = "DELETE FROM formulario_categoria WHERE categoriaId = ? AND formularioId = ?";
        this.ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        
        this.ps.setInt(1, categoria.getId());
        this.ps.setInt(2, formulario.getId());
        
        this.ps.executeUpdate();
        
        super.closeConnection();
    }
    
    @Override
    public void add(Formulario formulario, Categoria categoria) throws SQLException {
        super.openConnection();
        
        this.sql = "INSERT INTO formulario_categoria (formularioId, categoriaId) VALUES (?,?);";
        this.ps = connection.prepareStatement(this.sql, Statement.RETURN_GENERATED_KEYS);

        this.ps.setInt(1, formulario.getId());
        this.ps.setInt(2, categoria.getId());

        this.ps.executeUpdate();
        
        super.closeConnection();
    }
}

File: trunk/web/application/business/FormularioBusiness.php
<?php
require_once('Business.php');
require_once('CategoriaBusiness.php');
require_once(APPLICATION . '/config/IfspDatabase.php');
require_once(APPLICATION . '/models/Formulario.php');
require_once(APPLICATION . '/models/Categoria.php');

class FormularioBusiness extends Business {

    /**
     * Método para retornar formulário completo, contendo categorias e campos, pelo id
     * @param  int $id
     * @return Formulario formulario
     */

    public function obterFormularioCompleto($id) {
        parent::$database = IfspDatabase::getInstance();
        parent::$database->connect();

        $formulario = new Formulario();

        $query = "SELECT fc.formularioId, fc.categoriaId, fc.ordem "
               . "FROM formulario_categoria fc INNER JOIN formulario f ON fc.formularioId = f.id AND f.status = 1 AND f.aberto = 1 "
               . "INNER JOIN categoria c ON fc.categoriaId = c.id AND c.status = 1 "
               . "WHERE fc.formularioId = " . $id;

        parent::$database->query($query);

        // armazena resultados da query
        $resultados = array();
        while ($resultado = parent::$database->result()) {
            $resultados[] = $resultado;
        }

        // se existirem resultados
        if (count($resultados) > 0) {
            $categoriaBusiness = new CategoriaBusiness();
            $categorias = array();

            // recupera formulario
            $formulario = $this->find($id);

            // recupera categorias desse formulario
            foreach ($resultados as $resultado){
                $categoria = new Categoria();
                $categoria = $categoriaBusiness->obterCategoriaCompleta($resultado['categoriaId']);

                // adiciona a categoria a uma lista
                array_push($categorias, $categoria);
            }

            $formulario->setCategorias($categorias);
        }

        return $formulario;
    }

    /**
     * Método para retornar formulário pelo id
     * @param  int $id
     * @return Formulario formulario
     */
    public function find($id) {
        parent::$database = IfspDatabase::getInstance();
        parent::$database->connect();

        $formulario = new Formulario();

        $query = "SELECT f.id, f.titulo, f.aberto, f.status, f.data "
               . "FROM formulario f WHERE f.aberto = 1 AND f.status = 1 AND f.id = " . $id;

        parent::$database->query($query);

        if ($resultado = parent::$database->result()) {
            $this->assembly($formulario, $resultado);
        }

        return $formulario;
    }

    /**
     * Método para retornar formulários ativos
     * @return Formulario formulario
     */
    public function show() {
        parent::$database = IfspDatabase::getInstance();
        parent::$database->connect();

        $formularios = array();

        $query = "SELECT f.id, f.titulo, f.aberto, f.status, f.data "
               . "FROM formulario f WHERE f.aberto = 1 AND f.status = 1";

        parent::$database->query($query);

        while ($resultado = parent::$database->result()) {
            $formulario = new Formulario();
            $this->assembly($formulario, $resultado);

            array_push($formularios, $formulario);
        }

        return $formularios;
    }


    /**
     * Método para adicionar formulário ao banco
     * @param Model $model Formulario
     */
    public function add(Model $model) {
        throw new Exception("Método não implementado - Interface");
    }

    /**
     * Método para atualizar formulário no banco
     * @param Model $model Formulario
     */
    public function update(Model $model) {
        throw new Exception("Método não implementado - Interface");
    }

    public function __destruct() {
        if (isset(parent::$database) && parent::$database->getConnection()) {
            parent::$database->close();
        }
    }

    /**
     * Método para popular um Formulario
     * @param  Formulario $formulario
     * @param  row $resultado
     */
    private function assembly(Formulario $formulario, $resultado) {
        $formulario->setId($resultado['id']);
        $formulario->setTitulo($resultado['titulo']);
        $formulario->setAberto($resultado['aberto']);
        $formulario->setStatus($resultado['status']);
        $formulario->setData($resultado['data']);
    }
}
